File: libs/pi_hat/library/src/power_button.c
/**
 *
 * power_button.c
 *
 * Copyright (C) 2017 Kano Computing Ltd.
 * License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPL v2
 *
 * Power button interaction
 *
 */


#include <time.h>
#include <stdio.h>
#include <wiringPi.h>

#include "power_button.h"
#include "callbacks.h"
#include "err.h"
#include "pins.h"


const int button_timeout = 5;
callback_list *power_button_cbs;


int initialise_power_button()
{
    pullUpDnControl(POWER_PIN, DEFAULT_POWER_PIN_STATE);
    new_cb_list(&power_button_cbs);

    return SUCCESS;
}


int clean_up_power_button()
{
    free_cb_list(&power_button_cbs);

    return SUCCESS;
}


void press(void)
{
    static time_t last_call = 0;
    const time_t current_time = time(NULL);

    if (current_time - last_call > button_timeout) {
        last_call = current_time;
        dispatch_cbs(power_button_cbs);
    }
}


int register_power_off_cb(void cb(void))
{
    if (initialise() != SUCCESS)
        return E_FAIL;

    add_cb(power_button_cbs, cb);
    wiringPiISR(POWER_PIN, INT_EDGE_RISING, &press);

    return 0;
}

File: libs/pi_hat/examples/python/power_button.py
#!/usr/bin/env python
#
# power_button.py
#
# Copyright (C) 2017 Kano Computing Ltd.
# License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPLv2
#
# Interfacing with the power button through the raw library
#

import time

import paths
from kano_pi_hat.kano_hat import KanoHat


def power_pressed_cb():
    print 'Power button pressed'


if __name__ == '__main__':
    kano_hat = KanoHat()
    kano_hat.initialise()


    print "Hat is connected:", kano_hat.is_connected()
    kano_hat.register_power_off_cb(power_pressed_cb)

    while True:
        time.sleep(1)
